runtime/LibcWrappers.cpp

```diff
@@ -385,4 +385,298 @@ int SYM(memcmp)(const void *a, const void *b, size_t n) {
+    for (size_t i = 1; i < strlen(a); i++) {
+        ++aShadowIt;
+        ++bShadowIt;
+        allEqual =
+                _sym_build_bool_and(allEqual, _sym_build_equal(*aShadowIt, *bShadowIt));
+    }
+
+    _sym_push_path_constraint(allEqual, result == 0,
+                              reinterpret_cast<uintptr_t>(SYM(strcmp)));
+    return result;
+}
+
+int SYM(strncmp)(const char *a, const char *b, size_t n) {
+    tryAlternative(a, _sym_get_parameter_expression(0), SYM(strncmp));
+    tryAlternative(b, _sym_get_parameter_expression(1), SYM(strncmp));
+    tryAlternative(n, _sym_get_parameter_expression(2), SYM(strncmp));
+
+    auto result = strncmp(a, b, n);
+    _sym_set_return_expression(nullptr);
+
+    if (isConcrete(a, n) && isConcrete(b, n))
+        return result;
+
+    auto aShadowIt = ReadOnlyShadow(a, n).begin_non_null();
+    auto bShadowIt = ReadOnlyShadow(b, n).begin_non_null();
+    auto *allEqual = _sym_build_equal(*aShadowIt, *bShadowIt);
+    for (size_t i = 1; i < n; i++) {
+        ++aShadowIt;
+        ++bShadowIt;
+        allEqual =
+                _sym_build_bool_and(allEqual, _sym_build_equal(*aShadowIt, *bShadowIt));
+    }
+
+    _sym_push_path_constraint(allEqual, result == 0,
+                              reinterpret_cast<uintptr_t>(SYM(strncmp)));
+    return result;
+}
+
+uint32_t SYM(strlen)(const char *s) {
+    tryAlternative(s, _sym_get_parameter_expression(0), SYM(strlen));
+
+    // HACK! we regard strlen as a special strchr(s, '\0')
+    auto *result = strchr(s, 0);
+    _sym_set_return_expression(nullptr);
+
+    if (isConcrete(s, result != nullptr ? (result - s) : strlen(s)))
+        return (result - s);
+
+    // We force set the value of c to \x00, it should be a concrete value
+    auto *cExpr = _sym_build_integer(0, 8);
+
+    size_t length = result != nullptr ? (result - s) : strlen(s);
+    auto shadow = ReadOnlyShadow(s, length);
+    auto shadowIt = shadow.begin();
+    for (size_t i = 0; i < length; i++) {
+        _sym_push_path_constraint(
+                _sym_build_not_equal(
+                        (*shadowIt != nullptr) ? *shadowIt : _sym_build_integer(s[i], 8),
+                        cExpr),
+                /*taken*/ 1, reinterpret_cast<uintptr_t>(SYM(strchr)));
+        ++shadowIt;
+    }
+
+    // HACK! The last byte must be \x00!
+    _sym_push_path_constraint(
+            _sym_build_equal(
+                    (*shadowIt != nullptr) ? *shadowIt : _sym_build_integer(0, 8),
+                    cExpr),
+            /*taken*/ 1, reinterpret_cast<uintptr_t>(SYM(strchr)));
+
+    return (result - s);
+}
+
+int SYM(atoi)(const char *s) {
+    tryAlternative(s, _sym_get_parameter_expression(0), SYM(strlen));
+
+    // HACK! we mimic the libc function summary in klee
+    /* int atoi(const char *str) {
+     *     return (int)strtol(str, (char **)NULL, 10);
+     *   }
+     * */
+    auto result = strtol(s, (char **)NULL, 10);
```

Comment:
There's nothing wrong with that function summary, but why don't we call `atoi` directly?